pgmpy/factors/CPD.py

```diff
@@ -107,27 +107,25 @@ def __init__(self, variable, variable_card, values,
 
         cardinality = [variable_card]
         if evidence_card is not None:
-            if not isinstance(evidence_card, (list, set, tuple)):
+            if not isinstance(evidence_card, (list, tuple)):
                 if isinstance(evidence_card, np.ndarray):
                     evidence_card = evidence_card.tolist()
                 elif isinstance(evidence_card, (int, float)):
                     evidence_card = [evidence_card]
                 else:
-                    raise TypeError("Evidence cardinality must be list, set, "
+                    raise TypeError("Evidence cardinality must be list, "
                                     "tuple, numpy ndarray or simply a number.")
-            self.evidence_card = evidence_card
             cardinality.extend(evidence_card[::-1])
```

Comment:
@ankurankan I agree with you. So, I am not reversing the list finally and the output would be as in the third case. Also, in many other places, just like in `__str__` method, `evidence` was used directly, so the method implementation and their tests were designed that way only. I may need to look into those carefully and I will send in patches as I keep working on those to make sure new bugs are not introduced. 
In short, new implementation will be they way you suggested above.